debian/anoncoin.1
.TH ANONCOIN "1" "March 2014" "Anoncoin 0.8.5.6" "User Commands"
.SH NAME
anoncoin \- peer-to-peer network based anonymous digital currency
.SH DESCRIPTION
Anoncoin is a peer-to-peer
digital currency. Peer-to-peer (P2P) means that there is no central authority
to issue new money or keep track of transactions. Instead, these tasks are
managed collectively by the nodes of the network.
.SS "Usage:"
.PP
anoncoin [options]
.PP
anoncoin [options] <command> [params]  # Send command to \fB\-server\fR or anoncoind
.PP
anoncoin [options] help                # List commands
.PP
anoncoin [options] help <command>      # Get help for a command
.SH OPTIONS
.TP
\fB\-conf=\fR<file>
Specify configuration file (default: \fIanoncoin.conf\fR)
.TP
\fB\-pid=\fR<file>
Specify pid file (default: \fIanoncoind.pid\fR)
.TP
\fB-gen
Generate coins (default: \fI0\fR)
.TP
\fB\-datadir=\fR<dir>
Specify data directory
.TP
\fB\-dbcache=\fR<n>
Set database cache size in megabytes (default: \fI25\fR)
.TP
\fB\-timeout=\fR<n>
Specify connection timeout in milliseconds (default: \fI5000\fR)
.TP
\fB\-proxy=\fR<ip:port>
Connect through socks proxy
.TP
\fB\-socks=\fR<n>
Select the version of socks proxy to use (\fI4\fR\-\fI5\fR, default: \fI5\fR)
.TP
\fB\-tor=\fR<ip:port>
Use proxy to reach tor hidden services (default: same as \fB\-proxy\fR)
.TP
\fB\-dns\fR
Allow DNS lookups for \fB\-addnode\fR, \fB\-seednode\fR and \fB\-connect\fR
.TP
\fB\-port=\fR<port>
Listen for connections on <port> (default: \fI9377\fR or testnet: \fI19377\fR)
.TP
\fB\-maxconnections=\fR<n>
Maintain at most <n> connections to peers (default: \fI200\fR)
.TP
\fB\-addnode=\fR<ip>
Add a node to connect to and attempt to keep the connection open
.TP
\fB\-connect=\fR<ip>
Connect only to the specified node(s)
.TP
\fB\-seednode=\fR<ip>
Connect to a node to retrieve peer addresses, and disconnect
.TP
\fB\-externalip=\fR<ip>
Specify your own public address
.TP
\fB\-onlynet=\fR<net>
Only connect to nodes in network <net> (\fIIPv4\fR, \fIIPv6\fR, \fII2P\fR or \fITor\fR)
.TP
\fB\-discover\fR
Discover own IP address (default: \fI1\fR when listening and no \fB\-externalip\fR)
.TP
\fB\-checkpoints\fR
Only accept block chain matching built\-in checkpoints (default: \fI1\fR)
.TP
\fB\-listen\fR
Accept connections from outside (default: \fI1\fR if no \fB\-proxy\fR or \fB\-connect\fR)
.TP
\fB\-bind=\fR<addr>
Bind to given address and always listen on it. Use [host]:port notation for IPv6
.TP
\fB\-dnsseed\fR
Find peers using DNS lookup (default: \fI1\fR unless \fB\-connect\fR)
.TP
\fB\-irc\fR
Find peers using internet relay chat (default: \fI0\fR)
.TP
\fB\-banscore=\fR<n>
Threshold for disconnecting misbehaving peers (default: \fI100\fR)
.TP
\fB\-bantime=\fR<n>
Number of seconds to keep misbehaving peers from reconnecting (default: \fI86400\fR)
.TP
\fB\-maxreceivebuffer=\fR<n>
Maximum per\-connection receive buffer, <n>*1000 bytes (default: \fI5000\fR)
.TP
\fB\-maxsendbuffer=\fR<n>
Maximum per\-connection send buffer, <n>*1000 bytes (default: \fI1000\fR)
.TP
\fB\-bloomfilters\fR
Allow peers to set bloom filters (default: \fI1\fR)
.TP
\fB\-upnp\fR
Use UPnP to map the listening port (default: \fI1\fR when listening)
.TP
\fB\-paytxfee=\fR<amt>
Fee per KB to add to transactions you send
.TP
\fB\-mininput=\fR<amt>
When creating transactions, ignore inputs with value less than this (default: \fI0.0001\fR)
.TP
\fB\-daemon\fR
Run in the background as a daemon and accept commands
.TP
\fB\-testnet\fR
Use the test network
.TP
\fB\-debug\fR
Output extra debugging information. Implies all other \fB\-debug\fR* options
.TP
\fB\-debugnet\fR
Output extra network debugging information
.TP
\fB\-logtimestamps\fR
Prepend debug output with timestamp (default: \fI1\fR)
.TP
\fB\-shrinkdebugfile\fR
Shrink debug.log file on client startup (default: \fI1\fR when no \fB\-debug\fR)
.TP
\fB\-printtoconsole\fR
Send trace/debug info to console instead of debug.log file
.TP
\fB\-rpcuser=\fR<user>
Username for JSON\-RPC connections
.TP
\fB\-rpcpassword=\fR<pw>
Password for JSON\-RPC connections
.TP
\fB\-rpcport=\fR<port>
Listen for JSON\-RPC connections on <port> (default: \fI9376\fR or testnet: \fI19376\fR)
.TP
\fB\-rpcallowip=\fR<ip>
Allow JSON\-RPC connections from specified IP address
.TP
\fB\-rpcconnect=\fR<ip>
Send commands to node running on <ip> (default: \fI127.0.0.1\fR)
.TP
\fB\-rpcthreads=\fR<n>
Set the number of threads to service RPC calls (default: \fI4\fR)
.TP
\fB\-blocknotify=\fR<cmd>
Execute command when the best block changes (%s in cmd is replaced by block hash)
.TP
\fB\-walletnotify=\fR<cmd>
Execute command when a wallet transaction changes (%s in cmd is replaced by TxID)
.TP
\fB\-alertnotify=\fR<cmd>
Execute command when a relevant alert is received (%s in cmd is replaced by message)
.TP
\fB\-upgradewallet\fR
Upgrade wallet to latest format
.TP
\fB\-keypool=\fR<n>
Set key pool size to <n> (default: \fI100\fR)
.TP
\fB\-rescan\fR
Rescan the block chain for missing wallet transactions
.TP
\fB\-salvagewallet\fR
Attempt to recover private keys from a corrupt wallet.dat
.TP
\fB\-checkblocks=\fR<n>
How many blocks to check at startup (default: \fI288\fR, \fI0\fR = all)
.TP
\fB\-checklevel=\fR<n>
How thorough the block verification is (\fI0\fR\-\fI4\fR, default: \fI3\fR)
.TP
\fB\-txindex\fR
Maintain a full transaction index (default: \fI0\fR)
.TP
\fB\-loadblock=\fR<file>
Imports blocks from external blk000??.dat file
.TP
\fB\-reindex\fR
Rebuild block chain index from current blk000??.dat files
.TP
\fB\-par=\fR<n>
Set the number of script verification threads (up to \fI16\fR, \fI0\fR = auto, <\fI0\fR = leave that many cores free, default: \fI0\fR)
.SS "Block creation options:"
.TP
\fB\-blockminsize=\fR<n>
Set minimum block size in bytes (default: \fI0\fR)
.TP
\fB\-blockmaxsize=\fR<n>
Set maximum block size in bytes (default: \fI250000\fR)
.HP
\fB\-blockprioritysize=\fR<n> Set maximum size of high\-priority/low\-fee transactions in bytes (default: \fI27000\fR)
.SS "I2P Options:"
.TP
\fB\-generatei2pdestination\fR
Generate an I2P destination, print it and exit.
.TP
\fB-i2p\fR
Enable I2P
.TP
\fB\-onlynet\fR=\fIi2p\fR
Enable I2P only mode
.TP
\fB\-i2psessionname=\fR<session name>
Name of an I2P session. If it is not specified, value will be \fI'Anoncoin\-client'\fR
.TP
\fB\-samhost=\fR<ip or host name>
Address of the SAM bridge host. If it is not specified, value will be \fI'127.0.0.1'\fR.
.TP
\fB\-samport=\fR<port>
Port number of the SAM bridge host. If it is not specified, value will be \fI'7656'\fR.
.TP
\fB\-mydestination=\fR<pub+priv i2p\-keys>
Set full I2P destination (public+private keys). If it is not specified, the client will geneterate a random destination for you.
.SS "SSL options"
(see the Anoncoin Wiki for SSL setup instructions)
.TP
\fB\-rpcssl\fR
Use OpenSSL (https) for JSON\-RPC connections
.TP
\fB\-rpcsslcertificatechainfile=\fR<file.cert>
Server certificate file (default: \fIserver.cert\fR)
.TP
\fB\-rpcsslprivatekeyfile=\fR<file.pem>
Server private key (default: \fIserver.pem\fR)
.TP
\fB\-rpcsslciphers=\fR<ciphers>
Acceptable ciphers (default: \fITLSv1+HIGH:!SSLv2:!aNULL:!eNULL:!AH:!3DES:@STRENGTH\fR)
.SS "UI Options"
.TP
\fB-lang=\fR<lang>
Set language, for example \fI'de_DE'\fR (default: system locale)
.TP
\fB-min\fR
Start minimized
.TP
\fB-splash\fR
Show the splash screen on startup (default: \fI1\fR)
.SH "SEE ALSO"
\fBanoncoin.conf\fR(5), \fBanoncoind\fR(1)


src/qt/anoncoin.cpp
/*
 * W.J. van der Laan 2011-2012
 */
// Copyright 2013 The Anoncoin Developers
//
// I2P-patch
// Copyright (c) 2012-2013 giv

#include <QApplication>
#include <QMessageBox>
#include <QFile>
#include <QDir>
#include <QTextStream>

#include "bitcoingui.h"
#include "clientmodel.h"
#ifdef ENABLE_WALLET
#include "walletmodel.h"
#include "paymentserver.h"
#endif
#include "optionsmodel.h"
#include "guiutil.h"
#include "guiconstants.h"
#include "init.h"
#include "util.h"
#include "ui_interface.h"
#include "splashscreen.h"
#include "setupdarknet.h"

#include <QMessageBox>
#if QT_VERSION < 0x050000
#include <QTextCodec>
#endif
#include <QLocale>
#include <QTimer>
#include <QTranslator>
#include <QLibraryInfo>

#ifdef Q_OS_MAC
#include "macdockiconhandler.h"
#endif

#if defined(BITCOIN_NEED_QT_PLUGINS) && !defined(_BITCOIN_QT_PLUGINS_INCLUDED)
#define _BITCOIN_QT_PLUGINS_INCLUDED
#define __INSURE__
#include <QtPlugin>
Q_IMPORT_PLUGIN(qcncodecs)
Q_IMPORT_PLUGIN(qjpcodecs)
Q_IMPORT_PLUGIN(qtwcodecs)
Q_IMPORT_PLUGIN(qkrcodecs)
Q_IMPORT_PLUGIN(qtaccessiblewidgets)
#endif

// Declare meta types used for QMetaObject::invokeMethod
Q_DECLARE_METATYPE(bool*)

// Need a global reference for the notifications to find the GUI
static BitcoinGUI *guiref;
static SplashScreen *splashref;

static void ThreadSafeShowGeneratedI2PAddress(const std::string& caption, const std::string& pub, const std::string& priv, const std::string& b32, const std::string& configFileName)
{
    if(guiref)
    {
        QMetaObject::invokeMethod(guiref, "showGeneratedI2PAddr",
                                   GUIUtil::blockingGUIThreadConnection(),
                                   Q_ARG(QString, QString::fromStdString(caption)),
                                   Q_ARG(QString, QString::fromStdString(pub)),
                                   Q_ARG(QString, QString::fromStdString(priv)),
                                   Q_ARG(QString, QString::fromStdString(b32)),
                                   Q_ARG(QString, QString::fromStdString(configFileName)));
    }
    else
    {
        std::string msg = "\nIf you want to use a permanent I2P-address you have to set a \'mydestination\' option in the configuration file: ";
        msg += configFileName;
        msg += "\nGenerated address:\n";

        msg += "\nAddress + private key (save this text in the configuration file and keep it secret):\n";
        msg += priv;

        msg += "\n\nAddress (you can make it public):\n";
        msg += pub;

        msg += "\n\nShort base32-address:\n";
        msg += b32;
        msg += "\n\n";

        printf("%s: %s\n", caption.c_str(), msg.c_str());
        fprintf(stderr, "%s: %s\n", caption.c_str(), msg.c_str());
    }
}

static bool ThreadSafeMessageBox(const std::string& message, const std::string& caption, unsigned int style)
{
    // Message from network thread
    if(guiref)
    {
        bool modal = (style & CClientUIInterface::MODAL);
        bool ret = false;
        // In case of modal message, use blocking connection to wait for user to click a button
        QMetaObject::invokeMethod(guiref, "message",
                                   modal ? GUIUtil::blockingGUIThreadConnection() : Qt::QueuedConnection,
                                   Q_ARG(QString, QString::fromStdString(caption)),
                                   Q_ARG(QString, QString::fromStdString(message)),
                                   Q_ARG(unsigned int, style),
                                   Q_ARG(bool*, &ret));
        return ret;
    }
    else
    {
        printf("%s: %s\n", caption.c_str(), message.c_str());
        fprintf(stderr, "%s: %s\n", caption.c_str(), message.c_str());
        return false;
    }
}

static bool ThreadSafeAskFee(int64 nFeeRequired)
{
    if(!guiref)
        return false;
    if(nFeeRequired < CTransaction::nMinTxFee || nFeeRequired <= nTransactionFee || fDaemon)
        return true;

    bool payFee = false;

    QMetaObject::invokeMethod(guiref, "askFee", GUIUtil::blockingGUIThreadConnection(),
                               Q_ARG(qint64, nFeeRequired),
                               Q_ARG(bool*, &payFee));

    return payFee;
}

static void InitMessage(const std::string &message)
{
    if(splashref)
    {
        splashref->showMessage(QString::fromStdString(message), Qt::AlignBottom|Qt::AlignHCenter, QColor(55,55,55));
        qApp->processEvents();
    }
    printf("init message: %s\n", message.c_str());
}

/*
   Translate string to current locale using Qt.
 */
static std::string Translate(const char* psz)
{
    return QCoreApplication::translate("bitcoin-core", psz).toStdString();
}

/* Handle runaway exceptions. Shows a message box with the problem and quits the program.
 */
static void handleRunawayException(std::exception *e)
{
    PrintExceptionContinue(e, "Runaway exception");
    QMessageBox::critical(0, "Runaway exception", BitcoinGUI::tr("A fatal error occurred. Anoncoin can no longer continue safely and will quit.") + QString("\n\n") + QString::fromStdString(strMiscWarning));
    exit(1);
}

#ifndef BITCOIN_QT_TEST
int main(int argc, char *argv[])
{
    // Command-line options take precedence:
    ParseParameters(argc, argv);

#if QT_VERSION < 0x050000
    // Internal string conversion is all UTF-8
    QTextCodec::setCodecForTr(QTextCodec::codecForName("UTF-8"));
    QTextCodec::setCodecForCStrings(QTextCodec::codecForTr());
#endif

    Q_INIT_RESOURCE(bitcoin);
    QApplication app(argc, argv);

    // Register meta types used for QMetaObject::invokeMethod
    qRegisterMetaType< bool* >();

#ifdef ENABLE_WALLET
    // Do this early as we don't want to bother initializing if we are just calling IPC
    // ... but do it after creating app, so QCoreApplication::arguments is initialized:
    if (PaymentServer::ipcSendCommandLine())
        exit(0);
    PaymentServer* paymentServer = new PaymentServer(&app);
#endif

    // Install global event filter that makes sure that long tooltips can be word-wrapped
    app.installEventFilter(new GUIUtil::ToolTipToRichTextFilter(TOOLTIP_WRAP_THRESHOLD, &app));

    // ... then bitcoin.conf:
    if (!boost::filesystem::is_directory(GetDataDir(false)))
    {
        // This message can not be translated, as translation is not initialized yet
        // (which not yet possible because lang=XX can be overridden in bitcoin.conf in the data directory)
        QMessageBox::critical(0, "Anoncoin",
                              QString("Error: Specified data directory \"%1\" does not exist.").arg(QString::fromStdString(mapArgs["-datadir"])));
        return 1;
    }
    // Anoncoin
    if (!boost::filesystem::exists(GetConfigFile().string()))
    {
        // Run wizard
        runFirstRunWizard();
    }
    // Read config after it's potentional written by the wizard.
    ReadConfigFile(mapArgs, mapMultiArgs);

    // Application identification (must be set before OptionsModel is initialized,
    // as it is used to locate QSettings)
    QApplication::setOrganizationName("Anoncoin");
    QApplication::setOrganizationDomain("anoncoin.net");
    if(GetBoolArg("-testnet")) // Separate UI settings for testnet
        QApplication::setApplicationName("Anoncoin-Qt-testnet");
    else
        QApplication::setApplicationName("Anoncoin-Qt");

    // ... then GUI settings:
    OptionsModel optionsModel;

    // Get desired locale (e.g. "de_DE") from command line or use system locale
    QString lang_territory = QString::fromStdString(GetArg("-lang", QLocale::system().name().toStdString()));
    QString lang = lang_territory;
    // Convert to "de" only by truncating "_DE"
    lang.truncate(lang_territory.lastIndexOf('_'));

    QTranslator qtTranslatorBase, qtTranslator, translatorBase, translator;
    // Load language files for configured locale:
    // - First load the translator for the base language, without territory
    // - Then load the more specific locale translator

    // Load e.g. qt_de.qm
    if (qtTranslatorBase.load("qt_" + lang, QLibraryInfo::location(QLibraryInfo::TranslationsPath)))
        app.installTranslator(&qtTranslatorBase);

    // Load e.g. qt_de_DE.qm
    if (qtTranslator.load("qt_" + lang_territory, QLibraryInfo::location(QLibraryInfo::TranslationsPath)))
        app.installTranslator(&qtTranslator);

    // Load e.g. bitcoin_de.qm (shortcut "de" needs to be defined in bitcoin.qrc)
    if (translatorBase.load(lang, ":/translations/"))
        app.installTranslator(&translatorBase);

    // Load e.g. bitcoin_de_DE.qm (shortcut "de_DE" needs to be defined in bitcoin.qrc)
    if (translator.load(lang_territory, ":/translations/"))
        app.installTranslator(&translator);

    // Subscribe to global signals from core
    uiInterface.ThreadSafeMessageBox.connect(ThreadSafeMessageBox);
    uiInterface.ThreadSafeAskFee.connect(ThreadSafeAskFee);
    uiInterface.ThreadSafeShowGeneratedI2PAddress.connect(ThreadSafeShowGeneratedI2PAddress);
    uiInterface.InitMessage.connect(InitMessage);
    uiInterface.Translate.connect(Translate);

    // Show help message immediately after parsing command-line options (for "-lang") and setting locale,
    // but before showing splash screen.
    if (mapArgs.count("-?") || mapArgs.count("--help"))
    {
        GUIUtil::HelpMessageBox help;
        help.showOrPrint();
        return 1;
    }

#ifdef Q_OS_MAC
    // on mac, also change the icon now because it would look strange to have a testnet splash (green) and a std app icon (orange)
    if(GetBoolArg("-testnet")) {
        MacDockIconHandler::instance()->setIcon(QIcon(":icons/bitcoin_testnet"));
    }
#endif

    SplashScreen splash(QPixmap(), 0);
    if (GetBoolArg("-splash", true) && !GetBoolArg("-min"))
    {
        splash.show();
        splash.setAutoFillBackground(true);
        splashref = &splash;
    }

    if (mapArgs.count("-style"))
    {
        QString filename = QString::fromStdString((std::string)GetDataDir().string());
        filename = filename.append(QDir::separator()).append(QString::fromStdString((std::string)mapArgs["-style"]));
        QFile file(filename);
        if (file.exists())
        {
            if (file.open(QIODevice::ReadOnly | QIODevice::Text))
            {
                QTextStream in(&file);
                QString content = "";
                while (!in.atEnd())
                {
                    content.append(in.readLine());
                }
                app.setStyleSheet(content);
            }
            else
            {
                QMessageBox::warning(NULL, BitcoinGUI::tr("Failed to load style!"),
                    BitcoinGUI::tr("Failed to load the stylesheet provided."),
                    QMessageBox::Ok, QMessageBox::Ok);
            }
        }
    }

    app.processEvents();
    app.setQuitOnLastWindowClosed(false);

    try
    {
#ifndef Q_OS_MAC
        // Regenerate startup link, to fix links to old versions
        // OSX: makes no sense on mac and might also scan/mount external (and sleeping) volumes (can take up some secs)
        if (GUIUtil::GetStartOnSystemStartup())
            GUIUtil::SetStartOnSystemStartup(true);
#endif

        boost::thread_group threadGroup;

        BitcoinGUI window;
        guiref = &window;

        QTimer* pollShutdownTimer = new QTimer(guiref);
        QObject::connect(pollShutdownTimer, SIGNAL(timeout()), guiref, SLOT(detectShutdown()));
        pollShutdownTimer->start(200);

        if(AppInit2(threadGroup))
        {
            {
                // Put this in a block, so that the Model objects are cleaned up before
                // calling Shutdown().

                optionsModel.Upgrade(); // Must be done after AppInit2

                if (splashref)
                    splash.finish(&window);

                ClientModel clientModel(&optionsModel);
#ifdef ENABLE_WALLET
                WalletModel *walletModel = 0;
                if(pwalletMain)
                    walletModel = new WalletModel(pwalletMain, &optionsModel);
#endif
                window.setClientModel(&clientModel);
#ifdef ENABLE_WALLET
                if(walletModel)
                {
                    window.addWallet("~Default", walletModel);
                    window.setCurrentWallet("~Default");
                }
#endif

                // If -min option passed, start window minimized.
                if(GetBoolArg("-min"))
                {
                    window.showMinimized();
                }
                else
                {
                    window.show();
                }

#ifdef ENABLE_WALLET
                // Now that initialization/startup is done, process any command-line
                // bitcoin: URIs
                QObject::connect(paymentServer, SIGNAL(receivedURI(QString)), &window, SLOT(handleURI(QString)));
                QTimer::singleShot(100, paymentServer, SLOT(uiReady()));
#endif

                app.exec();

                window.hide();
                window.setClientModel(0);
#ifdef ENABLE_WALLET
                window.removeAllWallets();
#endif
                guiref = 0;
#ifdef ENABLE_WALLET
                delete walletModel;
#endif
            }
            // Shutdown the core and its threads, but don't exit Bitcoin-Qt here
            threadGroup.interrupt_all();
            threadGroup.join_all();
            Shutdown();
        }
        else
        {
            threadGroup.interrupt_all();
            threadGroup.join_all();
            Shutdown();
            return 1;
        }
    } catch (std::exception& e) {
        handleRunawayException(&e);
    } catch (...) {
        handleRunawayException(NULL);
    }
    return 0;
}
#endif // BITCOIN_QT_TEST
